objects/src/notes/vault.rs

```diff
@@ -0,0 +1,73 @@
+use super::{Asset, Digest, Hasher, NoteError, Vec, Word, WORD_SIZE};
+
+// NOTE VAULT
+// ================================================================================================
+/// An asset container for a note.
+///
+/// A note vault can contain up to 1000 assets. The entire vault can be reduced to a single hash
+/// which is computed by sequentially hashing the list of the vault's assets.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NoteVault {
+    assets: Vec<Asset>,
+    hash: Digest,
+}
+
+impl NoteVault {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    /// The maximum number of assets which can be carried by a single note.
+    pub const MAX_NUM_ASSETS: usize = 1000;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns an note asset vault constructed from the provided list of assets.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The list contains more than 1000 assets.
+    /// - There are duplicate assets in the list.
+    pub fn new(assets: &[Asset]) -> Result<Self, NoteError> {
+        if assets.len() > Self::MAX_NUM_ASSETS {
+            return Err(NoteError::too_many_assets(assets.len()));
+        }
+
+        let mut asset_elements = Vec::with_capacity(assets.len() * WORD_SIZE);
+
+        for (i, asset) in assets.iter().take(assets.len() - 1).enumerate() {
```

Comment:
Since we consume `assets` as iterator, why not take it as generic from the arguments?

Reply:
I couldn't figure out how to make it work with the next if statement without making the code much less readable.

Reply:
That's true, using iterators here might create unnecessary complexity